### PR TITLE
feat(property-filters): add not in cohort ability to filters

### PR DESCRIPTION
**Pull request by aspicer**
## Problem

Since switching to persons on events, we've been telling users to build cohorts to filter out internal users who may have changed their email.

However, we didn't offer them the ability to filter by "not in cohort" so they had to build a cohort of all their non internal users and then filter on that. That's much more inefficient.

## Changes

Add an operator to cohort and add the ability to filter by not in cohort.

image

image

image

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Tested in dev. Front-end changes mainly (not in cohort already implemented on backend).

### CODE REVIEW

File: frontend/src/lib/components/PropertyFilters/components/TaxonomicPropertyFilter.tsx [hunk 1]
```diff
@@ -65,7 +65,7 @@ export function TaxonomicPropertyFilter({
     ) => {
         selectItem(taxonomicGroup, value, item?.propertyFilterType)
         if (
-            taxonomicGroup.type === TaxonomicFilterGroupType.Cohorts ||
+            // taxonomicGroup.type === TaxonomicFilterGroupType.Cohorts ||
```
Comment (anirudhpillai) [inline frontend/src/lib/components/PropertyFilters/components/TaxonomicPropertyFilter.tsx:68]: guessing you meant to remove this

File: frontend/src/lib/utils.tsx [hunk 2]
```diff
@@ -219,6 +219,11 @@ export const selectorOperatorMap: Record<string, string> = {
     is_not: "≠ doesn't equal",
 }
 
+export const cohortOperatorMap: Record<string, string> = {
```
Comment (anirudhpillai) [inline frontend/src/lib/utils.tsx:222]: something like `user belongs to this cohort`/`user doesn't belong to this cohort`, might make it easier for users to understand what the selector does. Though will wait for Michael's input on this

File: frontend/src/types.ts [hunk 3]
```diff
@@ -633,6 +633,8 @@ export enum PropertyOperator {
     NotBetween = 'not_between',
     Minimum = 'min',
     Maximum = 'max',
+    In = 'in',
```
Comment (anirudhpillai) [inline frontend/src/types.ts:636]: dumb question but can't we reuse `IContains`/`NotIContains` here by putting them in the context of `PropertyFilterType.Cohort`? though also ok with current approach if you think it's cleaner :)

**Review by anirudhpillai — APPROVED** [review]
Code changes look good, though on UX side not fully sure if user's will understand the selector's actions

File: frontend/src/lib/utils.tsx (hunk 2 above)
Comment (aspicer) [reply frontend/src/lib/utils.tsx:222]: Yeah, the problem is how the operator component is structured on the frontend - the operator comes after the taxonomic selector. Having it move if you choose a cohort is not nice.

File: frontend/src/types.ts (hunk 3 above)
Comment (aspicer) [reply frontend/src/types.ts:636]: Isn't IContains a case insensitive string contains function? Seems like an overload.

**Review by Twixes — COMMENTED** [review]
Only looked at the UI: I _would_ say that the grammar is a bit unintuitive here.

"**where \<cohort\> contains users**" sounds like a filter on the cohort containing _any_ user. Singular would be less ambiguous: "**where \<cohort\> contains user**" – though still not the clearest way to convey the relation between the event and cohort. "**where user \<in/not in\> \<cohort\>**" sounds like the clearest option – but we'd need to break the informal rule of the taxonomic filter selector always being first.

Do with this as you will!

**Comment by aspicer** [comment]
Added "user in" / "user not in" and it seems good!
https://github.com/user-attachments/assets/4feb603c-d958-41aa-a6a5-43f40cacb434

**Pull request merged**